src/geode/geometry/rotation.cpp

```diff
@@ -0,0 +1,126 @@
+                      + quaternion.value( 2 ) * quaternion.value( 2 ) );
+        matrix[1][0] = 2.0
+                       * ( quaternion.value( 0 ) * quaternion.value( 1 )
+                           + quaternion.value( 2 ) * quaternion.value( 3 ) );
+        matrix[2][0] = 2.0
+                       * ( quaternion.value( 2 ) * quaternion.value( 0 )
+                           - quaternion.value( 1 ) * quaternion.value( 3 ) );
+        matrix[3][0] = 0.0;
+
+        matrix[0][1] = 2.0
+                       * ( quaternion.value( 0 ) * quaternion.value( 1 )
+                           - quaternion.value( 2 ) * quaternion.value( 3 ) );
+        matrix[1][1] =
+            1
+            - 2.0
+                  * ( quaternion.value( 2 ) * quaternion.value( 2 )
+                      + quaternion.value( 0 ) * quaternion.value( 0 ) );
+        matrix[2][1] = 2.0
+                       * ( quaternion.value( 1 ) * quaternion.value( 2 )
+                           + quaternion.value( 0 ) * quaternion.value( 3 ) );
+        matrix[3][1] = 0.0;
+
+        matrix[0][2] = 2.0
+                       * ( quaternion.value( 2 ) * quaternion.value( 0 )
+                           + quaternion.value( 1 ) * quaternion.value( 3 ) );
+        matrix[1][2] = 2.0
+                       * ( quaternion.value( 1 ) * quaternion.value( 2 )
+                           - quaternion.value( 0 ) * quaternion.value( 3 ) );
+        matrix[2][2] =
+            1
+            - 2.0
+                  * ( quaternion.value( 1 ) * quaternion.value( 1 )
+                      + quaternion.value( 0 ) * quaternion.value( 0 ) );
+        matrix[3][2] = 0.0;
+
+        matrix[0][3] = 0.0;
+        matrix[1][3] = 0.0;
+        matrix[2][3] = 0.0;
+        matrix[3][3] = 1.0;
+
+        return matrix;
+    }
+} // namespace
+
+namespace geode
+{
+    Vector3D opengeode_geometry_api rotate(
+        const Vector3D& vector, const Vector3D& axis, double angle )
+    {
+        if( axis.length2() < geode::global_epsilon2 )
+        {
+            OPENGEODE_EXCEPTION( axis.length2() < geode::global_epsilon2,
+                "[rotate] Given axis is almost null" );
+        }
+        while( angle < 0 )
+        {
+            angle += 2 * M_PI;
+        }
+        while( angle >= 2 * M_PI )
+        {
+            angle -= 2 * M_PI;
+        }
+
+        const auto base = axis.normalize() * std::sin( 0.5 * angle );
+        const geode::Vector< 4 > quaternion( { base.value( 0 ), base.value( 1 ),
+            base.value( 2 ), std::cos( 0.5 * angle ) } );
+
+        const auto matrix = rotation_matrix( quaternion );
+        const Vector< 4 > quat_vector(
+            { vector.value( 0 ), vector.value( 1 ), vector.value( 2 ), 1.0 } );
+        Vector< 4 > result;
+        for( const auto i : geode::Range{ 4 } )
+        {
+            double coord{ 0. };
+            for( const auto j : geode::Range{ 4 } )
+            {
+                coord += matrix[i][j] * quat_vector.value( j );
```

Comment:
Proposition: matrix could be defined as array<Vector<4>> and this operation could the resulat of a dot with the quaternion

Reply:
Good idea